### PR TITLE
Implement reinstall

### PR DESCRIPTION
**Pull request by kolsa**

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -83,20 +84,41 @@ func Run(spec *Specification) (*event.EventEmitter, error) {
...
 		Client:      client,
 		Network:     n,
 	}
 
+	hw, err := reg.ReadHardwareDetails()
+	if err != nil {
+		return eventEmitter, errors.Wrap(err, "unable to read all hardware details")
+	}
+
+	m, err := hammer.fetchMachine(spec.MachineUUID)
```
Comment (majst01) [inline cmd/root.go]: This is always nil if the machine was never registered at metal-api before,  and registering will not happen anymore ?

File: cmd/root.go [hunk 2]
```diff
@@ -108,12 +130,15 @@ func Run(spec *Specification) (*event.EventEmitter, error) {
 		if err != nil {
 			log.Warn("BIOS updates for this machine type are intentionally not supported, skipping EnsureUEFI", "error", err)
 		}
+	} else if firmware == "efi" {
+		log.Info("firmware is EFI")
+	} else {
+		log.Info("WHAT ELSE")
```
Comment (majst01) [inline cmd/root.go]: Why not logging the result of firmware ? This code is strange

File: cmd/root.go (hunk 1 above)
Comment (majst01) [reply cmd/root.go]: Am im wrong with my projection ?

Comment (kolsa) [reply cmd/root.go]: As with the current implementation it first reads the hardware details and returns if an error occurs. If this succeeds we fetch the machine and ignore any error, e.g. if the machine is unknown to metal-api, and proceed normally, i.e. wiping all disks, ...
In contrary, if we succeed to fetch a machine and this machine is allocated with a reinstall mark set, we skip the default workflow and proceed with the reinstallation process.

File: cmd/reinstall.go [hunk 3]
```diff
@@ -28,23 +28,26 @@ func (h *Hammer) fetchMachine(machineID string) (*models.ModelsV1MachineResponse
 
 // wipe only the disk that has the OS installed on one of its partitions, keep all other disks untouched
 func (h *Hammer) reinstall(m *models.ModelsV1MachineResponse, hw *models.DomainMetalHammerRegisterMachineRequest, eventEmitter *event.EventEmitter) (*kernel.Bootinfo, error) {
+	if m.Allocation.BootInfo == nil {
```
Comment (majst01) [inline cmd/reinstall.go]: can we check this already at metal-api to prevent reinstallation at all from the beginning ?

Comment (majst01) [reply cmd/reinstall.go]: This is good :+1:

File: cmd/register/register.go [hunk 4]
```diff
@@ -122,36 +115,50 @@ func (r *Register) readHardwareDetails() (*models.DomainMetalHammerRegisterMachi
...
 	if err != nil {
 		return nil, errors.Wrap(err, "unable to get system block devices")
 	}
-	disks := []*models.ModelsV1MachineBlockDevice{}
 	for _, disk := range blockInfo.Disks {
 		if strings.HasPrefix(disk.Name, storage.DiskPrefixToIgnore) {
 			continue
 		}
+		var parts []*models.ModelsV1MachineDiskPartition
+		containsOS := false // not yet allocated
```
Comment (majst01) [inline cmd/register/register.go]: containsOS is never set to false if i dont miss something

Comment (kolsa) [reply cmd/register/register.go]: You're right, this is an artifact of the previous implementation. We do not need this field at all. I'll remove this

**Pull request merged**